server/perkeepd/ui/blob_item_audio_content.js

```diff
@@ -0,0 +1,178 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+goog.provide('cam.BlobItemAudioContent');
+
+goog.require('goog.math.Size');
+
+// Renders audio blob items.
+cam.BlobItemAudioContent = React.createClass({
+	displayName: 'BlobItemAudioContent',
+
+	audioRef_: null,
+
+	propTypes: {
+		blobref: React.PropTypes.string.isRequired,
+		filename: React.PropTypes.string.isRequired,
+		href: React.PropTypes.string.isRequired,
+		size: React.PropTypes.instanceOf(goog.math.Size).isRequired,
+	},
+
+	getInitialState: function() {
+		return {
+			mouseover: false,
+			playing: false,
+		};
+	},
+
+	render: function() {
+		return React.DOM.div({
+				className: 'cam-blobitem-audio',
+				onMouseEnter: this.handleMouseOver_,
+				onMouseLeave: this.handleMouseOut_,
+			},
+			React.DOM.a({
+					className: 'cam-unstyled-button',
+					href: this.props.href
+				},
+				this.getAudio_(),
+				this.getPoster_()
+			),
+			this.getPlayPauseButton_(),
+		);
+	},
+
+	getPoster_: function() {
+		// TODO(oac): When server indexes audio and provides a poster image, render it here.
+		return React.DOM.i({
+			className: 'fa fa-volume-up',
+			style: {
+				fontSize: this.props.size.height / 1.5 + 'px',
+				lineHeight: this.props.size.height + 'px',
+				width: this.props.size.width,
+			},
+		});
+	},
+
+	getAudio_: function() {
+		return React.DOM.audio({
+			src: goog.string.subs('%s%s/%s', goog.global.CAMLISTORE_CONFIG.downloadHelper, this.props.blobref, this.props.filename),
+			controls: true,
+			ref: this.setAudioRef_,
+			style: {
+				display: 'none',
+			},
+		});
+	},
+
+	getPlayPauseButton_: function() {
+		if (!this.state.mouseover) {
+			return null;
+		}
+		return (
+			React.DOM.button({
+				className: classNames({
+					'cam-unstyled-button': true,
+					'cam-blobitem-audio-play': !this.state.playing,
+					'cam-blobitem-audio-pause': this.state.playing,
+				}),
+				onClick: this.handlePlayPauseClick_,
+			},
+				React.DOM.i({
+					className: classNames({
+						'fa': true,
+						'fa-play': !this.state.playing,
+						'fa-pause': this.state.playing,
+					}),
+					style: {
+						fontSize: this.props.size.height / 5 + 'px',
+					},
+				}),
+			)
+		);
+	},
+
+	setAudioRef_: function(audio) {
+		this.audioRef_ = audio;
+	},
+
+	handlePlayPauseClick_: function(e) {
+		if (this.state.playing) {
+			this.audioRef_.pause();
+			this.setState({
+				playing: false,
+			});
+		} else {
+			this.audioRef_.play();
+			this.setState({
+				playing: true,
+			});
+		}
+	},
+
+	handleMouseOver_: function() {
+		this.setState({
+			mouseover: true,
+		});
+	},
+
+	handleMouseOut_: function() {
+		this.setState({
+			mouseover: false,
+		});
+	},
+});
+
+cam.BlobItemAudioContent.isAudio = function(rm) {
+	// From https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats
+	var extensions = [
+		'webm',
+		'ogg',
```

Comment:
>>Good catch, I didn't know `ogg` is also a video format. Let's ensure video takes precedence
>
>Because you added the audio handler after the video one in BLOB_ITEM_HANDLERS_, I think it's already the case.
>
>>and use metadata to determine whether the blob is audio or video.
>
>yes, that would be nice.
>But on first approximation if you want, we can assume most oggs are in fact audio (which I believe is practically the case), and the video handler can always let the audio handler deal with oggs for now.

TODO: Give audio priority over video for `ogg`